### PR TITLE
Allow * in ACCEPT header

### PR DESCRIPTION
**Pull request by 82marbag**
## Motivation and Context
Allow `*/*`, `type/*` and `type/subtype`. This addresses https://github.com/awslabs/smithy-rs/issues/1544 and will comply with https://github.com/awslabs/smithy/issues/1332.

## Testing
`cargo test` in the Pokemon service and unit tests

## Checklist

- [x] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by Velfi — APPROVED** [review]

File: rust-runtime/aws-smithy-http-server/src/protocols.rs [hunk 1]
```diff
@@ -39,33 +39,75 @@ pub fn check_content_type<B>(
     }
 }
 
+pub fn accept_header_classifier<B>(req: &RequestParts<B>, content_type: &'static str) -> bool {
+    if req.headers().is_none() {
+        return true;
+    }
+    let headers = req.headers().unwrap();
+    if !headers.contains_key(http::header::ACCEPT) {
+        return true;
```
Comment (unexge) [inline rust-runtime/aws-smithy-http-server/src/protocols.rs:48]: Shouldn't we reject requests if there is no `Accept` header present?

Comment (82marbag) [reply rust-runtime/aws-smithy-http-server/src/protocols.rs:48]: From [here](https://httpwg.org/specs/rfc7231.html#header.accept),

> A request without any Accept header field implies that the user agent will accept any media type in response.

We're complying with this behavior here (same in the other implementations of smithy)

File: rust-runtime/aws-smithy-http-server/src/protocols.rs [hunk 2]
```diff
@@ -39,33 +39,75 @@ pub fn check_content_type<B>(
...
+                    let subtype = content_type.subtype();
+                    match (mim.type_(), mim.subtype()) {
+                        (t, s) if t == typ && s == subtype => true,
+                        (t, mime::STAR) if t == typ => true,
+                        (mime::STAR, mime::STAR) => true,
+                        _ => false,
+                    }
+                })
+                .unwrap_or(false)
+        })
```
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server/src/protocols.rs:76]: I think this piece of code deserve some extensive documentation. It's pretty hard to read, there is a lot going on between `map`, `flat_map`, multiple splits and so on.

Is the unwrap after the `nth(0)` safe?

**Review by crisidev — CHANGES_REQUESTED** [review]

File: rust-runtime/aws-smithy-http-server/src/protocols.rs [hunk 3]
```diff
@@ -39,33 +39,75 @@ pub fn check_content_type<B>(
...
+    if req.headers().is_none() {
+        return true;
+    }
+    let headers = req.headers().unwrap();
+    if !headers.contains_key(http::header::ACCEPT) {
+        return true;
+    }
+    let content_type = content_type
+        .parse::<mime::Mime>()
+        .expect("BUG: content_type is not valid");
```
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server/src/protocols.rs:52]:
```suggestion
        .expect("BUG: MIME parsing failed, content_type is not valid");
```

File: rust-runtime/aws-smithy-http-server/src/protocols.rs (hunk 2 above)
Comment (weihanglo) [reply rust-runtime/aws-smithy-http-server/src/protocols.rs]: > Is the unwrap after the nth(0) safe?

It should return an empty string for an empty string, so safe.

**Review by crisidev — APPROVED** [review]

**Pull request merged**